src/x3f_process.c

```diff
@@ -69,20 +59,57 @@ static int get_black_level(x3f_t *x3f,
   uint64_t *black_sum;
   double *black_sum_sqdev;
   int pixels, i;
+  char *cammodel;
+
+  char *name[4] = {"DarkShieldTop",
+		   "DarkShieldBottom",
+		   "FakeDarkShieldLeft",
+		   "FakeDarkShieldRight"};
+  int use[4] = {1, 1, 1, 1};
+  x3f_area16_t area[4];
+
+  if (image->channels < colors) return 0;
+
   /* Workaround for bug in DP2 firmware. DarkShieldBottom is specified
      incorrectly and thus ingnored. */
-  char *cammodel;
-  int use_bottom = !x3f_get_prop_entry(x3f, "CAMMODEL", &cammodel) ||
-    strcmp(cammodel, "SIGMA DP2");
+#define BOTTOM 1
+  use[BOTTOM] = (!x3f_get_prop_entry(x3f, "CAMMODEL", &cammodel) ||
+		 strcmp(cammodel, "SIGMA DP2"));
+
+  /* Real rects */
+  for (i=0; i<2; i++)
+    if (use[i])
+      use[i] = x3f_crop_area_camf(x3f, name[i], image, rescale, &area[i]);
+
+  /* Fake, column based rects */
+  for (i=2; i<4; i++)
+    if (use[i])
+      use[i] = x3f_crop_area_column(x3f, name[i], image, rescale, &area[i]);
+
+  for (i=0; i<4; i++)
+    if (use[i])
+      x3f_printf(DEBUG, "Calculate black level for %s\n", name[i]);
+    else
+      x3f_printf(DEBUG, "Do not calculate black level for %s\n", name[i]);
 
   pixels = 0;
   black_sum = alloca(colors*sizeof(uint64_t));
   memset(black_sum, 0, colors*sizeof(uint64_t));
-  pixels += sum_area(x3f, "DarkShieldTop", image, rescale, colors,
-		     black_sum);
-  if (use_bottom)
-    pixels += sum_area(x3f, "DarkShieldBottom", image, rescale, colors,
-		       black_sum);
+
+  x3f_printf(INFO, "Acc. dark level\n");
+  for (i=0; i<4; i++)
+    if (use[i]) {
+      int color;
+      pixels += sum_area(x3f, area[i], colors, black_sum);
+
+      x3f_printf(INFO, "  %s pixels = %d\n", name[i], pixels);
+
+      for (color = 0; color < colors; color++)
+	x3f_printf(INFO, "    mean[%d] = %f\n",
+		   color,
+		   (double)black_sum[color]/pixels);
```

Comment:
These printouts have to be removed before merging, or at least changed to DEBUG. If they are to be kept (as DEBUG), I also think they should be improved so that they print the mean for each area instead of cumulative means.